tests/mbt/model/ccv_pss.qnt

```diff
@@ -0,0 +1,157 @@
+    // and false otherwise.
+    pure def IsTopN(currentState: ProtocolState, validator: Node, consumer: Chain): bool = {
+        val proviValSet = currentState.providerState.chainState.currentValidatorSet
+        val N = currentState.providerState.topNByConsumer.get(consumer)
+
+        val topNValSet = GetTopNVals(proviValSet, N)
+
+        topNValSet.contains(validator)
+    }
+
+    // Returns true if the given validator has opted in to the given consumer chain,
+    pure def IsOptedIn(currentState: ProtocolState, validator: Node, consumer: Chain): bool = {
+        currentState.providerState.optedInVals.getOrElse(consumer, Set()).contains(validator)
+    }
+
+    // Opts a validator out. Safe to call before the consumer chain even runs.
+    // Will not stop the validator set from being forced to validate when in the top N.
+    // Validators that are in the top N will not be able to opt out, and
+    // an error will be returned.
+    // Similarly, if the validator is not opted in, an error will be returned.
+    pure def OptOut(currentState: ProtocolState, consumer: Chain, validator: Node): Result = {
+        if (currentState.IsTopN(validator, consumer)) {
+            Err("Cannot opt out a validator that is in the top N")
+        } else if (not(currentState.IsOptedIn(validator, consumer))) {
+            Err("Cannot opt out a validator that is not opted in")
+        } else {
+            pure val optedInVals = currentState.providerState.optedInVals.get(consumer)
+            pure val newOptedInVals = optedInVals.exclude(Set(validator))
+            Ok({
+                ...currentState,
+                providerState: {
+                    ...currentState.providerState,
+                    optedInVals: currentState.providerState.optedInVals.put(consumer, newOptedInVals)
+                }
+            })
+        }
+    }
+
+    // Runs the PSS logic that needs to run on endblock.
+    // Concretely, this will forcefully opt in all validators that are in the top N
+    // for each chain.
+    pure def endBlockPSS(providerState: ProviderState): ProviderState = {
+        val runningConsumers = providerState.getRunningConsumers()
+        runningConsumers.fold(
+            providerState,
+            (acc, consumer) => endBlockPSS_helper(acc, consumer)
+        )
+    }
+
+    // Runs the PSS logic for a single consumer.
+    // Should only be run for running chains.
+    pure def endBlockPSS_helper(providerState: ProviderState, consumer: Chain): ProviderState = {
+        val proviValSet = providerState.chainState.currentValidatorSet
+        val topNVals = GetTopNVals(proviValSet, providerState.topNByConsumer.get(consumer))
+        val prevOptedInVals = providerState.optedInVals.getOrElse(consumer, Set())
+        // opt in all the top N validators, i.e. union the top N vals with the previous opted in vals
+        val newOptedInVals = providerState.optedInVals.put(consumer, prevOptedInVals.union(topNVals))
```

Comment:
I guess because we have sets here, we're guaranteed that a validator that `IsTopN` and has opted in only appears once at the `optedInVals`, right?

Reply:
Yup, exactly!